### DISCUSSION TITLE
Error Pada Evaluasi Model (Confusion Matrix)

### DISCUSSION THREAD

**Discussion by joejnt** [teamalgoritma/community#512] [Q&A] [answered]
Dear rekan2 algoritma,

Saya mendapatkan error pada saat akan melakukan evaluasi model Decision Tree dengan menggunakan rpart.

**Splitting Dataset**
RNGkind(sample.kind = "Rounding")
set.seed(100)
intrainDT_Transaction <- initial_split(data = DT_Transaction,
                                       prop = 0.8,
                                       strata = "RISK_CATEGORY")
TrainDT_Transaction <- training(intrainDT_Transaction)
TestDT_Transaction <- testing(intrainDT_Transaction)

**Pemodelan**
DTrpart_transaction <- rpart(RISK_CATEGORY~., data = TrainDT_Transaction, method = "class")

**Evaluasi model**
cm_pred_dtree_rpart <- confusionMatrix(data = Pred_dtree_rpart , 
                                       reference = TrainDT_Transaction$RISK_CATEGORY)

Pada saat evaluasi model hasilnya error karena 

<img width="650" alt="image" src="https://user-images.githubusercontent.com/103427989/214512220-45349904-6623-46f6-9a02-499cea45c75c.png">

Terkait dengan pemodelan, apakah perbedaan menggunakan `rpart` dan `tree`? karena saya mendapatkan perbedaan pada saat pemodelan, mohon sarannya.

Best regards,
JT

**Comment by ramdhinafinita** [comment, marked as answer]
dear pak @joejnt 

Berdasarkan kode yang bapak berikan terjadi error dikarenakan:
seharusnya menggunakan  `TestDT_Transaction` pada fungsi predict() alih-alih menggunakan `TrainDT_Transaction` untuk mendapatkan prediksi yang baik.

Bapak dapat mencoba dengan menggunakan kode berikut:
> Pred_dtree_rpart <- predict(DTrpart_transaction, TestDT_Transaction , method ='class')
> Prob_dtree_rpart <- predict(DTrpart_transaction, TestDT_Transaction , method ='prob')

Selain itu, Bapak seharusnya menggunakan label set test dalam fungsi ConfusionMatrix(), alih-alih menggunakan label set train.

Bapak dapat menggunakan kode berikut:

> cm_pred_dtree_rpart <- confusionMatrix(data = Pred_dtree_rpart , reference = TestDT_Transaction$RISK_CATEGORY)

Silahkan untuk dicoba ya pak.

Selain itu, menjawab pertanyaan Bapak,
**rpart (recursive partitioning)** merupakan packages atau library untuk metode klasifikasi dan regresi.
Sedangkan **tree** merupakan sebuah fungsi base pada packages R yang dapat digunakan untuk klasifikasi dan regresi tree.
Kedua fungsi diatas sama-sama digunakan untuk membuat pohon keputusan atau decission tree model.

**Berikut merupakan perbedaan:**
`Syntax: `kode dalam membuat decssion tree model menggunakan rpart atau tree berbeda. Contoh, ketika kita menggunakan rpart, kita membutuhkan method argumen yang spesifik seperti "class" untuk  classification tree, atau "anova" untuk regression trees. Sedangkan jika menggunakan fungsi tree kita tidak ada argumen wajib yang perlu ditambahkan.

`Pruning/pemangkasan`: rpart memiliki built-in pruning mechanism, yang berarti dapat secara otomatis memutuskan cabang tree mana yang akan disimpan atau dihapus berdasarkan parameter klompeksifitas (cp). Sedangkan fungsi "tree" tidak punya mekanisme pemangkasan tree bawaan, sehingga sebagai gantinya kita harus memangkas tree secara manual dengan menggunakan packages/librari lain seperti "rpart.plot" untuk memvisualisasikan dan memangkas cabang pohin kita.

`Plotting`: rpart menyediakan cara sederhana untuk memplot pohon keputusan menggunakan fungsi plot(), tetapi tree tidak menyediakan fungsi plot bawaan apa pun.

`Speed`: Fungsi "tree" umumnya lebih cepat daripada rpart, terutama jika kumpulan datanya besar.

`Memory`: Fungsi "tree" lebih hemat memori daripada rpart, ia dapat menangani kumpulan data yang lebih besar daripada rpart.

Secara umum, rpart dan tree dapat digunakan untuk membangun model pohon keputusan /decission tree, tetapi rpart lebih cocok untuk kumpulan data besar dan memiliki mekanisme pemangkasan bawaan sementara pohon lebih cepat dan lebih hemat memori.

referensi lain: https://medium.com/@snehi.bhaskar26/what-is-the-difference-between-rpart-and-random-forest-package-15463fe74026

**Reply by joejnt** [reply]
Dear @ramdhinafinita 

Terima kasih untuk inside nya, namun pada syntax yang disarankan saya masih mendapatkan error.

<img width="623" alt="image" src="https://user-images.githubusercontent.com/103427989/214735734-f7de7b07-bf23-49f8-a2f9-68cde171e4ed.png">

mohon sarannya.

Best regards,
JT

**Reply by ramdhinafinita** [reply]
Dear Pak @joejnt 

Berdasarkan error bapak, error tersebut berasal dari parameter predict yang bapak pakai yakni "method" yang mana seharusnya "type".  Karena parameter method berada pada saat pembuatan model.

## Prediksi ke data train
Pred_dtree_rpart <- predict(object = DTrpart_transaction, newdata = TrainDT_Transaction , type='class')
## Prediksi ke data test
Prob_dtree_rpart <- predict(object = DTrpart_transaction, newdata = TestDT_Transaction , type='class')

## Confussion Matrix Data Train
confusionMatrix(data = Pred_dtree_rpart ,
                reference = TrainDT_Transaction$RISK_CATEGORY)

## Confussion Matrix Data Test
confusionMatrix(data = Prob_dtree_rpart ,
                reference = TestDT_Transaction$RISK_CATEGORY)

Silahkan dicoba ya pak.

**Reply by joejnt** [reply]
Dear @ramdhinafinita 

Well noted, terima kasih untuk insidenya.

Tambah lagi ya @ramdhinafinita, saya mau tanya tentang ploting data yang underfitting. Saya menjalankan syntax seperti ini:
<img width="617" alt="image" src="https://user-images.githubusercontent.com/103427989/214793779-91cdadab-6b86-4c4d-a885-32abc77c0685.png">

Dapat inside dari rpubs, tapi saya masih bingung bagaimana mengimpertasikannya, hasil plot nya seperti ini:
<img width="536" alt="image" src="https://user-images.githubusercontent.com/103427989/214793992-20556beb-4113-49af-bfff-362e4408d408.png">


Sebelumnya hasil pemodelan yang saya buat menggunakan ctree seperti ini:
<img width="326" alt="image" src="https://user-images.githubusercontent.com/103427989/214794143-e2d4fdb2-50bf-490b-9ad7-37bca9cf3c4d.png">

Awalnya saya pengen tau, apakah model yang saya buat underfit atau overfit.

Best regards,
JT

**Reply by ramdhinafinita** [reply]
Dear Pak @joejnt 

Izin menjawab ya pak, Saat kita menafsirkan plot pohon keputusan dengan parameter "underfitting" = TRUE, nilai model yang kita buat akan memiliki label kelas yang disusun dalam urutan tinggi ke rendah (High to Low). Ini berarti bahwa label kelas yang paling penting (yang memiliki nilai tertinggi) akan lebih dekat ke root dari tree nya, sedangkan label kelas yang paling tidak penting (yang memiliki nilai terendah) akan lebih jauh dari root. Ini berguna untuk mengidentifikasi label kelas mana yang paling penting dalam proses pengambilan keputusan dan mana yang kurang penting. Selain itu, ini juga dapat menunjukkan bahwa model mungkin tidak sesuai, artinya tidak cukup kompleks untuk secara akurat menangkap pola dalam data.

Sedangkan untuk menginterpretasi plot dari tree nya:
Untuk memahami apa yang ditampilkan pada pohon keputusan, kita mulai dari akar pohon/root of the tree (simpul keputusan paling pertama). Melihat simpul pertama, kita tahu bahwa variable yang paling menentukan keputusan adalah variable TOTAL_REKENING_DEPOSITO dari sekian macam variable yang digunakan pada model. Namun, jika kita lihat, hanya ada 3 variabel yang digunakan dalam simpul keputusan: yakni TOTAL_REKENING_DEPOSITO, TDBal_Nom, dan SALDO_TAB. Ini memberitahu kita bahwa ketiga variabel inilah yang paling penting ketika kita membedakan jenis kelas mana yang dimiliki pada setiap target (yakni kelas High ,Low, Low to Moderate, Moderate, dan Moderate to High)
Ada 3 leaves dari simpul ini: 
Contoh pertama yakni pada root dimana,
- Jika lebih dari 153 maka ditunjukkan oleh grafik pertama di sisi paling kanan, memberitahu kita bahwa ada 11 orang yang memiliki total rekening deposito lebih dari >153 dimana didalamnya terdapat 80% pada kelas High, dan 20% ada pada kelas Moderate to High. sumbu y pada grafik mewakili proposi total dalam kelas.
- Selanjutnya contoh untuk node kedua. variabel kedua yang menentukan keputusasn adalah TDBal_Nom dimana >50xxxx (cabang ke kanan) dan <= 50xxxx (cabang ke kiri). Jika >50xx 
terdapat 17 orang yang mana didominasi oleh kelas Moderate sekita 60%. dan seterusnya.

Tabel dibawah menunjukkan kelas nya dari High -> Low (diurutkan berdasarkan level) namun pada grafik decission tree ditampilkan berdasarkan abdjadnya. 

Sedangkan untuk melihat overfit atau Underfit dari data kita bisa melihat dari penurunan akurasi model, Kondisi Model:
- **Overfitting**: model baik di data train, namun amat buruk di data test
Presisi Data Train: 80%
Presisi Data Test: 50%

- **Underfitting**: model buruk di data train maupun di data test
Presisi Data Train: 40 %
Presisi Data Test: 30 %

- **Just Right**: model baik di data train, dan menurun sedikit (selisih < 0.1) -> yg kita inginkan
Presisi Data Train: 85 %
Presisi Data Test: 83% %

**Reply by joejnt** [reply]
Dear @ramdhinafinita 

Well noted, terima kasih untuk inside nya.

Best regards,
JT
